File: 70-tools/scripts/docs/regen-registry.py
#!/usr/bin/env python3
# ruff: noqa: E501,T201,S603,S607
"""
Regenerate `90-docs/_registry/docs.json` from disk front-matter.

The registry is a sidecar index used by documentation consumers and the
drift-check in `validate-adrs.py`. Markdown bodies remain canonical; this
script scans them and emits a normalized JSON table so a human or a CI
gate can see the full picture in one file.

Covered keys per entry:
    id, path, title, status, doc_type, topic, authoritative,
    authoritative_for, related, supersedes, superseded_by, amends,
    amended_by, last_verified

Idempotent. Sorts entries by `id`. Preserves existing non-ADR rows
(e.g. explanation / how-to / tutorial docs) — they're parsed the same
way.

Emits TWO sidecars from the same scan (kept in lock-step):
    docs.json — JSON table (snake_case string keys).
    docs.edn  — EDN table for babashka / Clojure consumers (kebab-case
                keyword keys, matching the repo's *.adr.edn convention).
                Read with `(clojure.edn/read-string (slurp ".../docs.edn"))`.

Usage:
    70-tools/scripts/docs/regen-registry.py
    70-tools/scripts/docs/regen-registry.py --check   # exits 1 on drift (json OR edn)
    70-tools/scripts/docs/regen-registry.py --json    # plan-only JSON (stdout)
    70-tools/scripts/docs/regen-registry.py --edn     # plan-only EDN (stdout)
"""

from __future__ import annotations

import argparse
import datetime
import json
import re
import sys
from pathlib import Path
from typing import Any

REPO = Path(__file__).resolve().parents[3]
REGISTRY = REPO / "90-docs" / "_registry" / "docs.json"
REGISTRY_EDN = REPO / "90-docs" / "_registry" / "docs.edn"
DOCS_ROOT = REPO / "90-docs"

# Keys we surface in the registry. Keep this list minimal and stable.
SURFACED_KEYS = (
    "id", "path", "title", "status", "doc_type", "topic",
    "authoritative", "authoritative_for", "related",
    "supersedes", "superseded_by", "amends", "amended_by",
    "depends_on",  # added cycle 64 for relation integrity coverage
    "last_verified",
)

# Valid status values. Anything else → validator flags it.
VALID_STATUSES = {"active", "proposed", "accepted", "deprecated", "superseded"}

# Valid doc_type values per 90-docs/CLAUDE.md.
VALID_DOC_TYPES = {"explanation", "reference", "how-to", "tutorial", "adr"}


def parse_frontmatter(text: str) -> dict[str, Any] | None:
    """Tiny YAML-ish front-matter parser (flat keys + list-of-scalars).

    Avoids a PyYAML dependency to keep the script hermetic. Doesn't
    handle nested mappings, but the repo's ADR frontmatters are flat
    with only list-of-strings at most.
    """
    if not text.startswith("---"):
        return None
    end = text.find("\n---", 3)
    if end < 0:
        return None
    block = text[3:end].strip("\n")
    out: dict[str, Any] = {}
    current_key = None
    for raw in block.splitlines():
        # strip inline comments conservatively (only after whitespace + '#')
        line = re.sub(r"\s+#.*$", "", raw)
        if not line.strip():
            current_key = None
            continue
        if line.startswith("  - ") or line.startswith("- "):
            item = line.lstrip(" -").strip().strip('"').strip("'")
            if current_key:
                out.setdefault(current_key, []).append(item)
            continue
        m = re.match(r"^([a-zA-Z_][a-zA-Z0-9_]*)\s*:\s*(.*)$", line)
        if not m:
            current_key = None
            continue
        k, v = m.group(1), m.group(2).strip()
        if v == "" or v == "[]":
            # Empty scalar opens a list (next lines may be `- item` entries);
            # `[]` closes the list immediately.
            out[k] = []
            current_key = None if v == "[]" else k
            continue
        # scalar with possible quotes
        stripped = v.strip('"').strip("'")
        if stripped.lower() in ("true", "false"):
            out[k] = stripped.lower() == "true"
        else:
            out[k] = stripped
        current_key = None
    return out


def scan_docs() -> list[dict[str, Any]]:
    entries: list[dict[str, Any]] = []
    for md in sorted(DOCS_ROOT.rglob("*.md")):
        # skip registry sidecars
        if "_registry" in md.parts:
            continue
        # skip CLAUDE.md — rule file, not a registered doc
        if md.name == "CLAUDE.md":
            continue
        try:
            text = md.read_text(encoding="utf-8")
        except Exception:
            continue
        fm = parse_frontmatter(text)
        if not fm or not fm.get("id"):
            continue
        rel = md.relative_to(REPO).as_posix()
        entry: dict[str, Any] = {"path": rel}
        for k in SURFACED_KEYS:
            if k == "path":
                continue
            if k in fm:
                entry[k] = fm[k]
        # Default shape defensiveness
        entry.setdefault("authoritative", False)
        for list_key in ("authoritative_for", "related", "supersedes",
                         "superseded_by", "amends", "amended_by"):
            if list_key not in entry or entry[list_key] is None:
                entry[list_key] = []
        entries.append(entry)
    entries.sort(key=lambda e: e.get("id", ""))
    return entries


def load_existing() -> dict[str, Any] | None:
    if not REGISTRY.exists():
        return None
    try:
        return json.loads(REGISTRY.read_text(encoding="utf-8"))
    except Exception:
        return None


def build_registry(entries: list[dict[str, Any]]) -> dict[str, Any]:
    return {
        "version": 2,
        "updated_at": datetime.date.today().isoformat(),
        "entries": entries,
    }


# ── EDN projection ─────────────────────────────────────────────────────────
# Faithful EDN of the SAME registry dict, for babashka / Clojure consumers.
# snake_case string keys → kebab-case keyword keys (repo *.adr.edn convention).
# Values stay as-is (strings/bools/ints/vectors); unicode is emitted verbatim.

def _edn_keyword(key: str) -> str:
    return ":" + key.replace("_", "-")


def _edn_string(s: str) -> str:
    out = (
        s.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\r", "\\r")
        .replace("\n", "\\n")
        .replace("\t", "\\t")
    )
    return f'"{out}"'


def _edn_value(v: Any) -> str:
    if v is None:
        return "nil"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, str):
        return _edn_string(v)
    if isinstance(v, list):
        return "[" + " ".join(_edn_value(x) for x in v) + "]"
    if isinstance(v, dict):
        return "{" + " ".join(f"{_edn_keyword(k)} {_edn_value(val)}" for k, val in v.items()) + "}"
    raise TypeError(f"un-encodable EDN value: {v!r}")


def render_edn(reg: dict[str, Any]) -> str:
    """One entry per line under :entries — diffs cleanly per-doc."""
    lines = [
        "{:version " + _edn_value(reg["version"]),
        " :updated-at " + _edn_value(reg["updated_at"]),
        " :entries",
    ]
    entries = reg["entries"]
    if not entries:
        lines.append(" []}")
        return "\n".join(lines) + "\n"
    body = "\n  ".join(_edn_value(e) for e in entries)
    lines.append(" [" + body + "]}")
    return "\n".join(lines) + "\n"


def _entries_json(entries: list[dict[str, Any]]) -> str:
    return json.dumps(entries, sort_keys=True, ensure_ascii=False)


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--check", action="store_true",
                    help="exit 1 if either on-disk sidecar (json or edn) differs from what we'd regenerate")
    ap.add_argument("--json", action="store_true",
                    help="emit the planned registry as JSON (stdout) instead of writing")
    ap.add_argument("--edn", action="store_true",
                    help="emit the planned registry as EDN (stdout) instead of writing")
    args = ap.parse_args()

    entries = scan_docs()
    new_reg = build_registry(entries)

    if args.json:
        print(json.dumps(new_reg, indent=2, ensure_ascii=False))
        return 0
    if args.edn:
        print(render_edn(new_reg), end="")
        return 0

    if args.check:
        # JSON sidecar — compare entries only (updated_at is volatile).
        old = load_existing() or {}
        old_entries = old.get("entries", [])
        if _entries_json(old_entries) != _entries_json(entries):
            print(f"registry drift detected: disk={len(entries)} entries, file={len(old_entries)} entries",
                  file=sys.stderr)
            print("run: 70-tools/scripts/docs/regen-registry.py", file=sys.stderr)
            return 1
        # EDN sidecar — compare the entries vector (ignore the volatile :updated-at line).
        want_edn = render_edn(new_reg)
        have_edn = REGISTRY_EDN.read_text(encoding="utf-8") if REGISTRY_EDN.exists() else ""

        def _edn_entries_block(text: str) -> str:
            idx = text.find(" :entries")
            return text[idx:] if idx != -1 else text

        if _edn_entries_block(have_edn) != _edn_entries_block(want_edn):
            print("registry EDN drift detected: docs.edn out of sync with docs.json source",
                  file=sys.stderr)
            print("run: 70-tools/scripts/docs/regen-registry.py", file=sys.stderr)
            return 1
        print(f"registry in sync ({len(entries)} entries; json + edn)")
        return 0

    # write both sidecars from the same scan
    REGISTRY.parent.mkdir(parents=True, exist_ok=True)
    REGISTRY.write_text(
        json.dumps(new_reg, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    REGISTRY_EDN.write_text(render_edn(new_reg), encoding="utf-8")
    print(f"wrote {REGISTRY.relative_to(REPO)} + {REGISTRY_EDN.relative_to(REPO)} with {len(entries)} entries")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: 70-tools/scripts/docs/regen-registry.clj
;; Regenerate 90-docs/_registry/{docs.json,docs.edn} from disk front-matter.
;;
;; babashka port of the former regen-registry.py (byte-for-byte identical
;; output — same scan order, same key order, same JSON pretty-format, same EDN).
;; clj-native per the repo's babashka direction; no PyYAML / Python dependency.
;;
;; The registry is a sidecar index used by documentation consumers and the
;; drift-check in validate-adrs.py. Markdown bodies remain canonical; this
;; script scans them and emits a normalized table so a human or a CI gate can
;; see the full picture in one file.
;;
;; Covered keys per entry:
;;   id, path, title, status, doc_type, topic, authoritative,
;;   authoritative_for, related, supersedes, superseded_by, amends,
;;   amended_by, depends_on, last_verified
;;
;; Idempotent. Sorts entries by `id`. Preserves existing non-ADR rows
;; (e.g. explanation / how-to / tutorial docs) — they're parsed the same way.
;;
;; Two sidecars from the same scan (kept in lock-step):
;;   docs.json — JSON table (snake_case string keys).
;;   docs.edn  — EDN table for babashka / Clojure consumers (kebab-case keyword
;;               keys, matching the repo's *.adr.edn convention).
;;
;; Usage:
;;   bb 70-tools/scripts/docs/regen-registry.clj
;;   bb 70-tools/scripts/docs/regen-registry.clj --check   ; exit 1 on drift (json OR edn)
;;   bb 70-tools/scripts/docs/regen-registry.clj --json    ; plan-only JSON (stdout)
;;   bb 70-tools/scripts/docs/regen-registry.clj --edn     ; plan-only EDN (stdout)

(ns regen-registry
  (:require [babashka.fs :as fs]
            [cheshire.core :as json]
            [clojure.edn :as edn]
            [clojure.string :as str]))

;; REPO = parents[3] of this script: docs -> scripts -> 70-tools -> REPO
(def repo
  (-> (or *file* "70-tools/scripts/docs/regen-registry.clj")
      fs/absolutize fs/parent fs/parent fs/parent fs/parent
      str))

(def docs-root (str (fs/path repo "90-docs")))
(def registry-json (str (fs/path repo "90-docs" "_registry" "docs.json")))
(def registry-edn (str (fs/path repo "90-docs" "_registry" "docs.edn")))

;; Non-path keys we surface, in the exact order the JSON entry dict is built.
(def surfaced-no-path
  ["id" "title" "status" "doc_type" "topic" "authoritative"
   "authoritative_for" "related" "supersedes" "superseded_by"
   "amends" "amended_by" "depends_on" "last_verified"])

(def list-default-keys
  ["authoritative_for" "related" "supersedes" "superseded_by" "amends" "amended_by"])

;; ── front-matter parser (faithful port of the Python "tiny YAML-ish") ───────

(defn- strip-set
  "Remove leading + trailing chars that are in `chars` (like Python str.strip(chars))."
  [s chars]
  (let [n (count s)
        start (loop [i 0] (if (and (< i n) (chars (.charAt s i))) (recur (inc i)) i))
        end   (loop [i n] (if (and (> i start) (chars (.charAt s (dec i)))) (recur (dec i)) i))]
    (subs s start end)))

(defn parse-frontmatter
  "Returns a map of front-matter keys, or nil if `text` has no `---` block."
  [text]
  (when (str/starts-with? text "---")
    (when-let [end (str/index-of text "\n---" 3)]
      (let [block (strip-set (subs text 3 end) #{\newline})
            lines (str/split block #"\n" -1)]
        (first
         (reduce
          (fn [[out ckey] raw]
            (let [line (str/replace raw #"\s+#.*$" "")]
              (cond
                (str/blank? line)
                [out nil]

                (or (str/starts-with? line "  - ") (str/starts-with? line "- "))
                (let [item (-> line
                               (str/replace #"^[ -]+" "")
                               str/trim
                               (strip-set #{\"})
                               (strip-set #{\'}))]
                  (if ckey
                    [(update out ckey (fnil conj []) item) ckey]
                    [out ckey]))

                :else
                (if-let [m (re-find #"^([a-zA-Z_][a-zA-Z0-9_]*)\s*:\s*(.*)$" line)]
                  (let [k (nth m 1)
                        v (str/trim (nth m 2))]
                    (if (or (= v "") (= v "[]"))
                      [(assoc out k []) (if (= v "[]") nil k)]
                      (let [stripped (-> v (strip-set #{\"}) (strip-set #{\'}))
                            lc (str/lower-case stripped)]
                        (if (or (= lc "true") (= lc "false"))
                          [(assoc out k (= lc "true")) nil]
                          [(assoc out k stripped) nil]))))
                  [out nil]))))
          [{} nil]
          lines))))))

;; ── ordered entry (a vector of [k v] pairs preserves Python insertion order) ─

(defn build-entry
  "Build one registry entry as an ordered [k v] pair-vector, mirroring the
  Python dict insertion order exactly."
  [rel fm]
  (let [base (into [["path" rel]]
                   (keep (fn [k] (when (contains? fm k) [k (get fm k)])) surfaced-no-path))
        base (if (some #(= "authoritative" (first %)) base)
               base
               (conj base ["authoritative" false]))
        base (reduce (fn [acc lk]
                       (if (some #(= lk (first %)) acc) acc (conj acc [lk []])))
                     base list-default-keys)]
    base))

(defn- entry-id [pairs]
  (or (some (fn [[k v]] (when (= k "id") v)) pairs) ""))

(defn- kw-name [k]
  (cond (keyword? k) (name k)
        (string? k) k
        :else (str k)))

(defn- scalar-str [v]
  (cond
    (keyword? v) (name v)
    (instance? java.util.Date v)
    (subs (str (java.time.Instant/ofEpochMilli (.getTime ^java.util.Date v))) 0 10)
    :else (str v)))

(defn- entity->fm
  "Extract registry front-matter map from an EDN tx-data entity
   ([{:db/id -1 :adr/id ...}] or bare map / :frontmatter form)."
  [m]
  (when (map? m)
    (let [base (if (map? (:frontmatter m))
                 (merge (dissoc m :frontmatter :body :body-file) (:frontmatter m))
                 m)
          ;; prefer :adr/* / :doc/* then bare keys
          getv (fn [& ks]
                 (some (fn [k]
                         (or (get base k)
                             (get base (keyword "adr" (name k)))
                             (get base (keyword "doc" (name k)))
                             (get base (name k))))
                       ks))
          id (getv :id :adr/id :doc/id)]
      (when id
        (cond-> {"id" (str id)
                 "title" (str (or (getv :title :adr/title :doc/title) id))
                 "status" (scalar-str (or (getv :status :adr/status :doc/status) "active"))
                 "doc_type" (scalar-str (or (getv :doc_type :doc-type :adr/doc_type :adr/doc-type :doc/doc_type) "doc"))
                 "topic" (scalar-str (or (getv :topic :adr/topic :doc/topic) ""))
                 "authoritative" (boolean (getv :authoritative :adr/authoritative :doc/authoritative))}
          (getv :last_verified :last-verified :adr/last_verified :adr/last-verified)
          (assoc "last_verified" (scalar-str (getv :last_verified :last-verified :adr/last_verified :adr/last-verified)))
          (sequential? (getv :related :adr/related :doc/related))
          (assoc "related" (mapv str (getv :related :adr/related :doc/related)))
          (sequential? (getv :authoritative_for :authoritative-for :adr/authoritative_for :adr/authoritative-for))
          (assoc "authoritative_for" (mapv str (getv :authoritative_for :authoritative-for :adr/authoritative_for :adr/authoritative-for)))
          (sequential? (getv :supersedes :adr/supersedes))
          (assoc "supersedes" (mapv str (getv :supersedes :adr/supersedes)))
          (sequential? (getv :superseded_by :superseded-by :adr/superseded_by :adr/superseded-by))
          (assoc "superseded_by" (mapv str (getv :superseded_by :superseded-by :adr/superseded_by :adr/superseded-by)))
          (sequential? (getv :amends :adr/amends))
          (assoc "amends" (mapv str (getv :amends :adr/amends)))
          (sequential? (getv :amended_by :amended-by :adr/amended_by :adr/amended-by))
          (assoc "amended_by" (mapv str (getv :amended_by :amended-by :adr/amended_by :adr/amended-by))))))))

(defn- parse-edn-doc
  "Parse EDN doc file → front-matter map or nil."
  [text]
  (try
    (let [form (edn/read-string {:default (fn [_tag v] v)} text)]
      (cond
        (and (vector? form) (map? (first form))) (entity->fm (first form))
        (map? form) (or (entity->fm form)
                        (when-let [fm (parse-frontmatter (or (:body form) ""))]
                          (when (get fm "id") fm)))
        :else nil))
    (catch Exception _ nil)))

(defn scan-docs []
  ;; EDN-only SSoT (ADR-2607171600 / superproject alignment): scan **.edn and
  ;; legacy **.md / **.md.edn. Prefer .edn when both exist.
  (let [md-files (->> (fs/glob docs-root "**.md") (map str))
        md-edn-files (->> (fs/glob docs-root "**.md.edn") (map str))
        edn-files (->> (fs/glob docs-root "**.edn")
                       (map str)
                       (remove #(str/ends-with? % ".md.edn"))
                       (remove #(str/includes? % "/_registry/"))
                       (remove #(= "deps.edn" (fs/file-name %))))
        ;; path without extension for dedupe: prefer edn > md.edn > md
        strip (fn [p]
                (-> p
                    (str/replace #"\.md\.edn$" "")
                    (str/replace #"\.edn$" "")
                    (str/replace #"\.md$" "")))
        by-base (atom {})]
    (doseq [p md-files]
      (swap! by-base assoc (strip p) p))
    (doseq [p md-edn-files]
      (swap! by-base assoc (strip p) p))
    (doseq [p edn-files]
      (swap! by-base assoc (strip p) p))
    (->> (vals @by-base)
         sort
         (keep (fn [p]
                 (let [parts (set (map str (fs/components p)))
                       name (fs/file-name p)]
                   (when-not (or (contains? parts "_registry")
                                 (= "CLAUDE.md" name)
                                 (= "CLAUDE.edn" name)
                                 (= "deps.edn" name))
                     (let [text (try (slurp p) (catch Exception _ nil))
                           fm (when text
                                (cond
                                  (str/ends-with? p ".edn") (or (parse-edn-doc text)
                                                                (parse-frontmatter text))
                                  :else (parse-frontmatter text)))]
                       (when (and fm (get fm "id"))
                         (build-entry (str (fs/relativize repo p)) fm)))))))
         (sort-by entry-id)
         vec)))

(defn today-iso [] (str (java.time.LocalDate/now)))

;; ── JSON encoder — byte-identical to json.dumps(obj, indent=2, ensure_ascii=False) ──

(defn- json-escape [s]
  (let [sb (StringBuilder.)]
    (doseq [c s]
      (let [code (int c)]
        (cond
          (= c \\) (.append sb "\\\\")
          (= c \") (.append sb "\\\"")
          (= code 8)  (.append sb "\\b")
          (= code 12) (.append sb "\\f")
          (= code 10) (.append sb "\\n")
          (= code 13) (.append sb "\\r")
          (= code 9)  (.append sb "\\t")
          (< code 32) (.append sb (format "\\u%04x" code))
          :else (.append sb c))))
    (str sb)))

(defn- json-str [s] (str \" (json-escape s) \"))

(declare json-encode)

(defn- json-pairs [pairs level]
  (if (empty? pairs)
    "{}"
    (let [pad  (apply str (repeat (* 2 level) \space))
          pad2 (apply str (repeat (* 2 (inc level)) \space))
          items (map (fn [[k v]] (str pad2 (json-str k) ": " (json-encode v (inc level)))) pairs)]
      (str "{\n" (str/join ",\n" items) "\n" pad "}"))))

(defn- json-array [xs level]
  (if (empty? xs)
    "[]"
    (let [pad  (apply str (repeat (* 2 level) \space))
          pad2 (apply str (repeat (* 2 (inc level)) \space))
          items (map (fn [x] (str pad2 (json-encode x (inc level)))) xs)]
      (str "[\n" (str/join ",\n" items) "\n" pad "]"))))

(defn json-encode [v level]
  (cond
    (string? v)  (json-str v)
    (boolean? v) (if v "true" "false")
    (integer? v) (str v)
    (vector? v)  (json-array v level)
    :else (throw (ex-info "un-encodable JSON value" {:v v}))))

(defn render-json [entries date]
  ;; top-level object: version / updated_at / entries (entries = vector of pair-vectors)
  (let [pad2 "  "
        ev (if (empty? entries)
             "[]"
             (str "[\n"
                  (str/join ",\n" (map #(str "    " (json-pairs % 2)) entries))
                  "\n  ]"))]
    (str "{\n"
         pad2 "\"version\": 2,\n"
         pad2 "\"updated_at\": " (json-str date) ",\n"
         pad2 "\"entries\": " ev "\n"
         "}\n")))

;; ── EDN encoder — byte-identical to the Python render_edn ───────────────────

(defn- edn-keyword [k] (str ":" (str/replace k "_" "-")))

(defn- edn-string [s]
  (str \"
       (-> s
           (str/replace "\\" "\\\\")
           (str/replace "\"" "\\\"")
           (str/replace "\r" "\\r")
           (str/replace "\n" "\\n")
           (str/replace "\t" "\\t"))
       \"))

(defn- edn-value [v]
  (cond
    (string? v)  (edn-string v)
    (boolean? v) (if v "true" "false")
    (integer? v) (str v)
    (vector? v)  (str "[" (str/join " " (map edn-value v)) "]")
    :else (throw (ex-info "un-encodable EDN value" {:v v}))))

(defn- edn-entry [pairs]
  (str "{" (str/join " " (map (fn [[k v]] (str (edn-keyword k) " " (edn-value v))) pairs)) "}"))

(defn render-edn [entries date]
  (let [head (str "{:version 2\n :updated-at " (edn-string date) "\n :entries")]
    (if (empty? entries)
      (str head "\n []}\n")
      (str head "\n [" (str/join "\n  " (map edn-entry entries)) "]}\n"))))

;; ── modes ───────────────────────────────────────────────────────────────────

(defn- pairs->map [pairs] (into {} pairs))

(defn- edn-entries-block [text]
  (let [idx (str/index-of text " :entries")]
    (if idx (subs text idx) text)))

(defn -main [& args]
  (let [args (set args)
        entries (scan-docs)
        date (today-iso)]
    (cond
      (contains? args "--json")
      (do (print (render-json entries date)) 0)

      (contains? args "--edn")
      (do (print (render-edn entries date)) 0)

      (contains? args "--check")
      (let [;; JSON sidecar — compare entries data only (updated_at is volatile)
            disk (try (json/parse-string (slurp registry-json) false) (catch Exception _ {}))
            disk-entries (get disk "entries" [])
            gen-entries (mapv pairs->map entries)]
        (cond
          (not= disk-entries gen-entries)
          (do (binding [*out* *err*]
                (println (str "registry drift detected: disk=" (count entries)
                              " entries, file=" (count disk-entries) " entries"))
                (println "run: nbb scripts/run-task.cljs docs:registry"))
              1)

          (not= (edn-entries-block (if (fs/exists? registry-edn) (slurp registry-edn) ""))
                (edn-entries-block (render-edn entries date)))
          (do (binding [*out* *err*]
                (println "registry EDN drift detected: docs.edn out of sync with docs.json source")
                (println "run: nbb scripts/run-task.cljs docs:registry"))
              1)

          :else
          (do (println (str "registry in sync (" (count entries) " entries; json + edn)")) 0)))

      :else
      (do
        (fs/create-dirs (fs/parent registry-json))
        (spit registry-json (render-json entries date))
        (spit registry-edn (render-edn entries date))
        (println (str "wrote 90-docs/_registry/docs.json + 90-docs/_registry/docs.edn with "
                      (count entries) " entries"))
        0))))

;; Run main unless a LIBRARY CONSUMER has said it only wants the functions. The guard used to be
;; `(= *file* (System/getProperty "babashka.file"))`, which asks "am I babashka's entry script?" --
;; false under every other runtime, so running this generator with `clojure` loaded it, regenerated
;; nothing and exited 0. The polarity was backwards: the default should be "I am the program",
;; because that is the case a silent no-op ruins, and the one caller that load-file's this (see
;; test_regen_registry.clj) is the one that knows it wants a library and can say so.
(when-not (System/getProperty "regen-registry.library-load")
  (System/exit (apply -main *command-line-args*)))
